llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp

```diff
@@ -493,6 +494,103 @@ class ValueToShadowMap {
   DenseMap<Value *, Value *> Map;
 };
 
+// Base class for handling some details of __nsan_* functions
+class NsanInstrumentationFunction {
+public:
+  virtual std::pair<FunctionCallee, int>
+  getCalleeAndNumArgsForMemoryOp(const uint64_t MemOpSize) const = 0;
+  virtual FunctionCallee getGeneralFunction() const = 0;
+  virtual ~NsanInstrumentationFunction() = default;
+};
+
+// Helper class for __nsan_copy_* family
```

Comment:
How about simplifying / removing duplication, i.e. like this:
```

template <size_t N>
class MemOp {
public:
    MemOp(Module &M, ArrayRef<StringRef> Sized, StringRef Fallback);

    FunctionCallee getFunctionFor(MemIntrinsic &MI) const;    
    FunctionCallee getFallback() const;   
private:
    std::array<FunctionCallee, N> functions;
};

MemOp<4> NsanCopy;
MemOp<4> NsanSetUnknown;
```

Reply:
thanks for the tip, removing duplication is really something that would be nice here

Reply:
Done. I use the second template argument to pass information about number of arguments of the functions represented by the instance

Reply:
Or we can just pass this number as constructor argument?